### ISSUE TITLE
Conv3D/Conv2d performance on Tesla arch. (half precision)

### ISSUE THREAD

**Issue by rrkarim** [pytorch/pytorch#24082]
## 🐛 Bug

I'm just testing the new release and observe the increased time consumption for the basic `conv2d/conv3d` layers. I'm more interested in `float16` performance. Here is the simple script:

``` 
import time
import torch
import torch.nn as nn

dtype = torch.float32
device = torch.device("cuda:0")

mod = nn.Conv3d(64, 64, 3).to(device, dtype=dtype)
input_ = torch.randn((8, 64, 64, 64, 64), device=device, dtype=dtype)

curr = time.time()
torch.cuda.synchronize()

for i in range(100):
    mod(input_)

torch.cuda.synchronize()
print(time.time() - curr)
```

It takes `4.002` for `float32` and `6.214` for `float16`. 

UPD: testing now with the newest `CUDA toolkit`, will include the results here.
UPD2: Tested on `py3.6_cuda10.0.130_cudnn7.6.2_0`, the results are same.  

 - PyTorch Version (e.g., 1.2.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`): 
 - CUDA/cuDNN version: 7.6.0
 - GPU models and configuration: RTX 2080 Ti

cc @ezyang @gchanan @zou3519

**Comment by soumith** [comment]
cc: @ptrblck @csarofeen 

I wonder if it is nothing more than a CuDNN upgrade screwing up perf (it wont be the first time ;-) ). In this release we are using CuDNN 7.6.2.

@rrkarim can you please try with `torch.backends.cudnn.benchmark=True`? That will tell us whether it's clearly CuDNN, or if we have to start digging elsewhere.

**Comment by ptrblck** [comment]
@rrkarim Could you swap these lines:
```python
curr = time.time()
torch.cuda.synchronize()
# to
torch.cuda.synchronize()
curr = time.time()
```
as you are currently also timing the initialization of the layers?

If you are running with `torch.backends.cudnn.benchmark = True`, could you please add some warmup iterations (10-50) before starting the real profiling?

**Comment by rrkarim** [comment]
@ptrblck @soumith it is the same.

```
import time
import torch
import torch.nn as nn

torch.backends.cudnn.benchmark=True

dtype = torch.float32
device = torch.device("cuda:0")

mod = nn.Conv3d(64, 64, 3).to(device, dtype=dtype)
input_ = torch.randn((8, 64, 64, 64, 64), device=device, dtype=dtype)

for i in range(40):
    mod(input_)

torch.cuda.synchronize()
curr = time.time()

for i in range(100):
    mod(input_)

torch.cuda.synchronize()
print(time.time() - curr)

```

**Comment by rrkarim** [comment]
Setting kernel to one (`mm`) I can witness increase in performance.

**Comment by ptrblck** [comment]
I've profiled the convolution using [this code](https://gist.github.com/ptrblck/cd38dbdd0570741aa44f2725ccbc1288).

Results:

FP32:

|         | cudnn=True, benchmark=True | cudnn=True, benchmark=False | cudnn=False |
|---------|----------------------------|-----------------------------|-------------|
| V100    | 3.355                      | 3.364                       | 6.198       |
| Titan V | 3.551                      | 3.550                       | 6.512       |
| P100    | 5.483                      | 5.477                       | 8.966       |

FP16:

|         | cudnn=True, benchmark=True | cudnn=True, benchmark=False | cudnn=False |
|---------|----------------------------|-----------------------------|-------------|
| V100    | 3.516                      | 3.523                       | 3.200       |
| Titan V | 4.089                      | 4.092                       | 3.217       |
| P100    | 5.456                      | 5.456                       | 12.132      |

While the FP32 results look reasonable, the FP16 results look indeed fishy.
We'll perform some internal tests and try to debug this issue.

**Comment by ptrblck** [comment]
@rrkarim Could you provide some additional workloads (shapes) which would be important for your use cases?

**Comment by rrkarim** [comment]
@ptrblck not sure about the use case I was trying to work on, since I have posted the issue long time ago and switched to other problems. But the example script above (first message of the thread) is pretty good use case for me. I don't have an access to a Turing arch. for now, therefore I can not test the performance. If everything is solved, you can close the issue.

**Comment by ptrblck** [comment]
This issue should be solved in cudnn `7.6.5` as some heuristics were improved and we see these numbers on a V100 for FP16 now:

|         | cudnn=True, benchmark=True | cudnn=True, benchmark=False | cudnn=False |
|---------|----------------------------|-----------------------------|-------------|
| V100    | 0.7892                      | 0.7894                       | 2.9869       |
| Titan V | 0.8612                      | 0.8611                       | 3.4629       |

@rrkarim You could build from source using the latest cudnn release to get these improvements.
Also, please let me know, if you see any issues.

**Comment by rrkarim** [comment]
@ptrblck thanks, I will check it ASAP.

**Comment by ngimel** [comment]
Closing, feel free to reopen if the issue persists.
